### ISSUE TITLE
Certain recursive scipy.integrate.quad (e.g. dblquad and nquad) give incorrect results when compiled with ifort

### ISSUE THREAD

**Issue by michiel88** [scipy/scipy#6882]
I noticed that a pre-compiled scipy for windows from http://www.lfd.uci.edu/~gohlke/pythonlibs/ gave incorrect results for certain recursive calls for integrals with singularities. See the small example on the bottom. Also after compiling it myself the problem persisted.

The problem is that quadpack is F77 code, which is not suppose to be called recursively. I think ifort uses static storage for some of its local arrays. Compiling quadpack with the /recursive flag solves this problem. 

In the end I have changed $PYTHON\Lib\site-packages\numpy\distutils\fcompiler\intel.py:162 from
opt = ['/nologo', '/MD', '/nbs', '/names:lowercase', '/assume:underscore'] to
opt = ['/nologo', '/MD', '/nbs', '/names:lowercase', '/assume:underscore', '/recursive']

My Debian version seems to be unaffected since the /openmp flag is used by default (both /recursive and /openmp activate the required /auto option).

Note that this is also related to ~~http://stackoverflow.com/questions/30913664/scipy-integrate-quad-gives-wrong-result-on-large-ranges~~ http://stackoverflow.com/questions/27270044/big-integration-error-with-integrate-nquad and https://mail.scipy.org/pipermail/scipy-dev/2014-December/020239.html

```python
from numpy import sqrt
from scipy.integrate import quad

def integrand(x, y):
    return (1.0 / sqrt(x) - 1.0) * (1.0 / sqrt(y) - 1.0) * sqrt(x + y)

def inner_int(y):
    return quad(integrand, 0, 1, args=(y))[0]

a = quad(inner_int, 0, 1)
print(a)
# Correct: (0.5214054330650386, 1.9894822456123507e-09)
# Incorrect: (1.455338119269806, 1.550319872478667e-10)
# Reference: https://www.wolframalpha.com/input/?i=integrate+(1.0+%2F+sqrt(x)+-+1.0)+*+(1.0+%2F+sqrt(y)+-+1.0)+*+sqrt(x+%2B+y)+dx+dy,+x%3D0..1,+y+%3D+0..1
```

**Comment by ev-br** [comment]
Cannot repro with gcc/gfortran on linux:

```
In [5]: quad(inner_int, 0, 1)
Out[5]: (0.5214054330650386, 1.9894822456123507e-09)

In [6]: import scipy

In [7]: scipy.__version__
Out[7]: '0.18.1'
```

same answer on scipy 0.17.1 and -dev.

**Comment by pv** [comment]
gfortran allocates "small" arrays on stack, so it's recursion-proof in small problem cases.
You can try with smaller -fmax-stack-var-size=1

Lapack iirc works around this (it is also a threadsafety issue) by setting specific compiler flags. Of course, numpy.distutils makes this more difficult, but we should consider monkeypatching or upstreaming the fix.

**Comment by cgohlke** [comment]
I have updated the binaries at http://www.lfd.uci.edu/~gohlke/pythonlibs/#scipy. The `/recursive` flag worked for me.

**Comment by siddpiku** [comment]
I am getting the incorrect answer in anaconda python 3.5 in windows 10.  May sound silly, but can someone tell me how to recompile quadpack in anaconda after I make the changes

**Comment by rgommers** [comment]
@siddpiku that's not easy for an extension with Fortran code on Windows. If you have access to the Intel Fortran compiler, see https://software.intel.com/en-us/articles/numpyscipy-with-intel-mkl. Otherwise, I'd advise you not to try.

**Comment by siddpiku** [comment]
@rgommers - That looks difficult to me. Do you know any other way of making the integration posted above (or [here](http://scicomp.stackexchange.com/questions/26168/numerical-double-integration-with-endpoint-singularity-in-scipy-python-gives-inc) ) give correct answer with anaconda for windows?

**Comment by cgohlke** [comment]
> Do you know any other way of making the integration posted above (or here ) give correct answer with anaconda for windows?

You could try to uninstall scipy from Anaconda and `pip install scipy-0.18.1-cp35-cp35m-win_amd64.whl` from http://www.lfd.uci.edu/%7Egohlke/pythonlibs/#scipy. Or just replace the file `_quadpack.cp35-win_amd64.pyd` in your Anaconda installation with the file contained in `scipy-0.18.1-cp35-cp35m-win_amd64.whl` (which is a ZIP file). Chances are it will work but I have not tried.

**Comment by siddpiku** [comment]
@cgohlke thanks that worked! Though I had to uninstall and re-install both numpy and scipy from  http://www.lfd.uci.edu/%7Egohlke/pythonlibs/, since numpy didnot havelink to Intel mkl libraries. I was wondering, maybe somone should fix this in the libraries installed through anaconda. Where can File a bug report?

**Comment by pv** [comment]
Now that we have dropped g77 support, it would be simple to sprinkle 
`recursive subroutine/function` Fortran 90 definitions in the sources 
everywhere appropriate.

**Comment by WarrenWeckesser** [comment]
@pv wrote:

> Now that we have dropped g77 support, it would be simple to sprinkle `recursive subroutine/function` Fortran 90 definitions in the sources everywhere appropriate.

Does anyone have a Windows development environment in which to try this?  Or can we get a test case that would currently fail on AppVeyor, and then abuse the CI by hacking on the Fortran until the test passes on AppVeyor?

**Comment by WarrenWeckesser** [comment]
I gave it a shot in https://github.com/scipy/scipy/pull/9469, but I suspect that if I had looked a little more closely at above comments and how Appveyor builds scipy, I would have know that the tests would pass.

**Comment by mdhaber** [comment]
@rgommers I've tested this on Windows with Anaconda Python and I get the correct result. Can we close this or is something else needed?

**Comment by rgommers** [comment]
@mdhaber with the Intel Fortran compiler?

**Comment by mdhaber** [comment]
Nope, my mistake.

**Comment by ev-br** [comment]
gh-15997 provides a mechanical fix. Would be nice to check with ifort whether that patch actually fixes the problem. Which I sadly cannot do ATM, so would be great if somebody can.

**Comment by ev-br** [comment]
Ok, so gh-15997 is merged (1.9.0-dev), let's close this. Please reopen if it causes problems. Thanks all
